local h = require("../../helpers")
local Signal = require("../../../luks-std/Signal")

return function(expect)
    h.suite("Signal Basic Creation", function()
        h.test("Signal.new returns a valid signal", function()
            local sig = Signal.new()
            expect.not_nil(sig)
            expect.type_(sig.Connect, "function")
            expect.type_(sig.Fire, "function")
            expect.type_(sig.Public, "function")
        end)

        h.test("Public proxy is frozen", function()
            local sig = Signal.new()
            local proxy = sig:Public()
            expect.true_(table.isfrozen(proxy))
        end)

        h.test("Public proxy exposes only allowed methods", function()
            local sig = Signal.new()
            local proxy = sig:Public()
            expect.not_nil(proxy.Connect)
            expect.not_nil(proxy.Once)
            expect.not_nil(proxy.Wait)
            expect.nil_(proxy.Fire)
            expect.nil_(proxy.DisconnectAll)
            expect.nil_(proxy.Public)
        end)
    end)
end
